### PR TITLE
Improved messages for json errors

### PR DESCRIPTION
**Pull request by YohDeadfall**
Use error messages that better reflect
- https://doc.rust-lang.org/std/option/enum.Option.html#recommended-message-style
- https://rustc-dev-guide.rust-lang.org/diagnostics.html

### CODE REVIEW

File: pgrx/src/datum/json.rs [hunk 1]
```diff
@@ -44,7 +44,7 @@ impl FromDatum for Json {
             let len = varsize_any_exhdr(varlena);
             let data = vardata_any(varlena);
             let slice = std::slice::from_raw_parts(data as *const u8, len);
-            let value = serde_json::from_slice(slice).expect("failed to parse Json value");
+            let value = serde_json::from_slice(slice).expect("failed to parse a json value");
```
Comment (workingjubilee) [inline pgrx/src/datum/json.rs]: The reason it's called `expect` is that it should encode the _expectation_ instead of _what went wrong_ (because what went wrong is imperfect knowledge, because we often aren't checking and doesn't help the programmer... we want to give a _why_). We've been casual about this, but if we're going to spend time revising these:

Let's rephrase this as "Datum must refer to a JSON varlena" or something similar (if we've failed at parsing JSON, then it's not JSON, according to us).

File: pgrx/src/datum/json.rs [hunk 2]
```diff
@@ -67,12 +67,12 @@ impl FromDatum for JsonB {
                 pg_sys::jsonb_out,
                 &[Some(detoasted.into())],
             )
-            .expect("failed to convert jsonb to a cstring");
+            .expect("failed to convert a jsonb to a cstring");
```
Comment (workingjubilee) [inline pgrx/src/datum/json.rs]: Likewise, "Datum must refer to JSONB for Postgres to serialize it to a cstring" or something similar.

File: pgrx/src/datum/json.rs [hunk 3]
```diff
@@ -67,12 +67,12 @@ impl FromDatum for JsonB {
...
 
             let value = serde_json::from_str(
-                cstr.to_str().expect("text version of jsonb is not valid UTF8"),
+                cstr.to_str().expect("a text version of the jsonb is not valid UTF8"),
             )
-            .expect("failed to parse JsonB value");
+            .expect("failed to parse a jsonb value");
```
Comment (workingjubilee) [inline pgrx/src/datum/json.rs]: reframe in terms of requirements, etc.

File: pgrx/src/datum/json.rs [hunk 4]
```diff
@@ -122,7 +122,7 @@ impl FromDatum for JsonString {
 /// for json
 impl IntoDatum for Json {
     fn into_datum(self) -> Option<pg_sys::Datum> {
-        let string = serde_json::to_string(&self.0).expect("failed to serialize Json value");
+        let string = serde_json::to_string(&self.0).expect("failed to serialize a json value");
```
Comment (workingjubilee) [inline pgrx/src/datum/json.rs]: reframe in terms of requirements, etc.

File: pgrx/src/datum/json.rs [hunk 5]
```diff
@@ -134,9 +134,9 @@ impl IntoDatum for Json {
 /// for jsonb
 impl IntoDatum for JsonB {
     fn into_datum(self) -> Option<pg_sys::Datum> {
-        let string = serde_json::to_string(&self.0).expect("failed to serialize JsonB value");
-        let cstring =
-            alloc::ffi::CString::new(string).expect("string version of jsonb is not valid UTF8");
+        let string = serde_json::to_string(&self.0).expect("failed to serialize a jsonb value");
+        let cstring = alloc::ffi::CString::new(string)
+            .expect("string version of jsonb contains a non-terminating nul");
```
Comment (workingjubilee) [inline pgrx/src/datum/json.rs]: reframe in terms of requirements, etc.

**Review by workingjubilee — CHANGES_REQUESTED** [review]

File: pgrx/src/datum/json.rs (hunk 1 above)
Comment (YohDeadfall) [reply pgrx/src/datum/json.rs]: A follow up question then. Are the other guidelines then how to format messages there? In the standard library I saw that everything is lowercase and in some other crates too.

Comment (workingjubilee) [reply pgrx/src/datum/json.rs]: That's because of https://rustc-dev-guide.rust-lang.org/diagnostics.html?highlight=style#diagnostic-output-style-guide

Comment (YohDeadfall) [reply pgrx/src/datum/json.rs]: Should `pgrx` follow it?

Comment (workingjubilee) [reply pgrx/src/datum/json.rs]: It's usually good style to follow the stdlib, but I'm not that pressed over doing so.

File: pgrx/src/datum/json.rs (hunk 4 above)
Comment (YohDeadfall) [reply pgrx/src/datum/json.rs]: An attempt to find a correct wording has been failed completely. The thing here is that `serde_json` should be able to serialize it's own types without any issue. Moreover, the `Value` type says that it's a valid JSON representation.

File: pgrx/src/datum/json.rs [hunk 6]
```diff
@@ -67,12 +68,12 @@ impl FromDatum for JsonB {
                 pg_sys::jsonb_out,
                 &[Some(detoasted.into())],
             )
-            .expect("failed to convert jsonb to a cstring");
+            .expect("Ddtum must refer to a valid JSONB varlena");
```
Comment (SteveLauC) [inline pgrx/src/datum/json.rs]:
```suggestion
            .expect("Datum must refer to a valid JSONB varlena");
```

A tiny typo:>

Comment (YohDeadfall) [reply pgrx/src/datum/json.rs]: Thanks! Fixed and lower-cased.

File: pgrx/src/datum/json.rs (hunk 4 above)
Comment (workingjubilee) [reply pgrx/src/datum/json.rs]: Ah. Then if we don't have anything useful to say, simply `.unwrap()`.

File: pgrx/src/datum/json.rs [hunk 7]
```diff
@@ -67,12 +68,12 @@ impl FromDatum for JsonB {
                 pg_sys::jsonb_out,
                 &[Some(detoasted.into())],
             )
-            .expect("failed to convert jsonb to a cstring");
+            .expect("datum must refer to a valid jsonb varlena");
 
             let value = serde_json::from_str(
-                cstr.to_str().expect("text version of jsonb is not valid UTF8"),
+                cstr.to_str().expect("a text version of the jsonb must be a valid utf-8"),
```
Comment (workingjubilee) [inline pgrx/src/datum/json.rs]:
```suggestion
                cstr.to_str().expect("a text version of the jsonb must be valid utf-8"),
```

File: pgrx/src/datum/json.rs [hunk 8]
```diff
@@ -67,12 +68,12 @@ impl FromDatum for JsonB {
...
             )
-            .expect("failed to parse JsonB value");
+            .expect("a text version of jsonb must be a vaild json");
```
Comment (workingjubilee) [inline pgrx/src/datum/json.rs]:
```suggestion
            .expect("a text version of jsonb must be a valid json");
```

**Review by workingjubilee — APPROVED** [review]
thank you for trying. this looks good with the nits addressed.
